client/src/main/java/org/asynchttpclient/cookie/ThreadSafeCookieStore.java

```diff
@@ -0,0 +1,264 @@
+        return Objects.equals(cookiePath, requestPath) ||
+                (requestPath.startsWith(cookiePath) && (cookiePath.charAt(cookiePath.length() - 1) == '/' || requestPath.charAt(cookiePath.length()) == '/'));
+    }
+
+    private static class CookieKey implements Comparable<CookieKey> {
+        final String name;
+        final String domain;
+        final String path;
+
+        CookieKey(String name, String domain, String path) {
+            this.name = name;
+            this.domain = domain;
+            this.path = path;
+        }
+
+        @Override
+        public int compareTo(CookieKey o) {
+            int result;
+            if (o != null) {
+                if ((result = this.name.compareTo(o.name)) == 0)
+                    if ((result = this.domain.compareTo(o.domain)) == 0)
+                        result = this.path.compareTo(o.path);
+            } else
+                throw new NullPointerException();
+
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            return obj instanceof CookieKey && this.compareTo((CookieKey) obj) == 0;
+        }
+
+        @Override
+        public int hashCode() {
+            int result = 17;
+            result = 31 * result + name.hashCode();
+            result = 31 * result + domain.hashCode();
+            result = 31 * result + path.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s: %s; %s", name, domain, path);
+        }
+    }
+
+    private static class StoredCookie {
+        final Cookie cookie;
+        final boolean hostOnly;
+        final boolean persistent;
+        final long createdAt = System.currentTimeMillis();
+
+        StoredCookie(Cookie cookie, boolean hostOnly, boolean persistent) {
+            this.cookie = cookie;
+            this.hostOnly = hostOnly;
+            this.persistent = persistent;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s; hostOnly %s; persistent %s", cookie.toString(), hostOnly, persistent);
+        }
+    }
+
+    private void add(String requestDomain, String requestPath, Cookie cookie) {
+
+        AbstractMap.SimpleEntry<String, Boolean> pair = cookieDomain(cookie.domain(), requestDomain);
+        String keyDomain = pair.getKey();
+        boolean hostOnly = pair.getValue();
+        String keyPath = cookiePath(cookie.path(), requestPath);
+        CookieKey key = new CookieKey(cookie.name().toLowerCase(), keyDomain, keyPath);
+
+        if (hasCookieExpired(cookie, 0))
+            cookieJar.remove(key);
+        else
+            cookieJar.put(key, new StoredCookie(cookie, hostOnly, cookie.maxAge() != Cookie.UNDEFINED_MAX_AGE));
+    }
+
+    private List<Cookie> get(String domain, String path, boolean secure) {
+
+        final boolean[] removeExpired = {false};
```

Comment:
why an array?

Reply:
Limitation of Java8 lambdas.

Reply:
Damn. OK. I guess a array of booleans of size 1 is smaller than an AtomicBoolean.